src/apply.c

```diff
@@ -480,3 +483,110 @@ int git_apply_tree(
 
 	return error;
 }
+
+/* normal: apply to workdir: ignore index
+ * --cached: apply to index: ignore workdir
+ * --index: apply to both: validate index == workdir
+ */
+
+int git_apply(
+	git_repository *repo,
+	git_diff *diff,
+	git_apply_options *given_opts)
+{
+	git_index *contents = NULL, *repo_index = NULL;
+	git_reader *pre_reader = NULL;
+	const git_diff_delta *delta;
+	git_vector paths = GIT_VECTOR_INIT;
+	git_apply_options opts = GIT_APPLY_OPTIONS_INIT;
+	git_checkout_options checkout_opts = GIT_CHECKOUT_OPTIONS_INIT;
+	bool do_checkout;
+	size_t i;
+	int error;
+
+	assert(repo && diff);
+
+	GITERR_CHECK_VERSION(
+		given_opts, GIT_APPLY_OPTIONS_VERSION, "git_apply_options");
+
+	if (given_opts)
+		memcpy(&opts, given_opts, sizeof(git_apply_options));
+
+	do_checkout = (opts.location != GIT_APPLY_LOCATION_INDEX);
+
+	/*
+	 * by default, we apply a patch directly to the working directory;
+	 * in `--cached` or `--index` mode, we apply to the contents already
+	 * in the index.
+	 */
+	if (opts.location == GIT_APPLY_LOCATION_WORKDIR)
+		error = git_reader_for_workdir(&pre_reader, repo);
+	else
+		error = git_reader_for_index(&pre_reader, repo, NULL);
+
+	if (error < 0)
+		goto done;
+
+	/* if we're not checking out, we're writing to the repo's index */
+	if (do_checkout)
+		error = git_vector_init(&paths, git_diff_num_deltas(diff), NULL);
+	else
+		error = git_repository_index(&repo_index, repo);
+
+	if (error < 0)
+		goto done;
+
+	/*
+	 * note: this is not the full postimage, this only contains the
+	 * new files created during the diffing process.  we will limit
+	 * checkout to only write the files affected by this diff.
+	 */
+	if ((error = git_index_new(&contents)) < 0)
+		goto done;
+
+	for (i = 0; i < git_diff_num_deltas(diff); i++) {
+		delta = git_diff_get_delta(diff, i);
+
+		if ((error = apply_one(repo, pre_reader, contents, diff, i)) < 0)
+			goto done;
+
+		if (do_checkout) {
+			git_vector_insert(&paths, (void *)delta->old_file.path);
+
+			if (strcmp(delta->old_file.path, delta->new_file.path))
```

Comment:
I'm slightly concerned that those `{old,new}_file` members could be `NULL` ?

Reply:
They cannot - generated diffs [always have `old_file.path`](https://github.com/libgit2/libgit2/blob/master/src/diff_generate.c#L47) and [`new file.path`](https://github.com/libgit2/libgit2/blob/master/src/diff_generate.c#L53) set.  And parsed patches [ensure that they're both set](https://github.com/libgit2/libgit2/blob/master/src/patch_parse.c#L972).